### PR TITLE
[SYCLomatic] Support migration of 14 cusolver API

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/test/dpct/helper_files_ref/include/lapack_utils.hpp [hunk 1]
```diff
@@ -909,6 +909,123 @@ oneapi::mkl::compz job2compz(const oneapi::mkl::job &job) {
...
+  void operator()(sycl::queue &q, oneapi::mkl::job jobz, oneapi::mkl::uplo uplo,
+                  std::int64_t n, library_data_t a_type, void *a,
+                  std::int64_t lda, void *w, void *device_ws,
+                  std::size_t device_ws_size, int *info) {
+    using value_t = typename value_type_trait<T>::value_type;
+    auto a_data = dpct::detail::get_memory(reinterpret_cast<T *>(a));
+    auto device_ws_data =
+        dpct::detail::get_memory(reinterpret_cast<T *>(device_ws));
+    auto w_data = dpct::detail::get_memory(reinterpret_cast<value_t *>(w));
+    if constexpr (std::is_same_v<T, float> || std::is_same_v<T, double>) {
```
Comment (Alexander-Kleymenov) [inline clang/test/dpct/helper_files_ref/include/lapack_utils.hpp]: JFYI, it's also possible to use is_floating_point to check this condition.

**Review by Alexander-Kleymenov — COMMENTED** [review]
Looks good to me.

File: clang/test/dpct/helper_files_ref/include/lapack_utils.hpp (hunk 1 above)
Comment (zhiweij1) [reply clang/test/dpct/helper_files_ref/include/lapack_utils.hpp]: OK, I will update it. Thanks for the review.

File: clang/runtime/dpct-rt/include/lapack_utils.hpp.inc [hunk 2]
```diff
@@ -1148,6 +1148,158 @@ oneapi::mkl::compz job2compz(const oneapi::mkl::job &job) {
...
+template <typename T> struct syheev_scratchpad_size_impl {
+  void operator()(sycl::queue &q, oneapi::mkl::compz jobz,
+                  oneapi::mkl::uplo uplo, std::int64_t n, std::int64_t lda,
+                  std::size_t &device_ws_size) {
+#ifndef __INTEL_MKL__
+    throw std::runtime_error(
+        "The oneAPI Math Kernel Library (oneMKL) Interfaces "
+        "Project does not support this API.");
+#else
+    if constexpr (std::is_floating_point_v<T>) {
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/lapack_utils.hpp.inc]: 1166 - 1172 can be a macro

Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/lapack_utils.hpp.inc]: done

**Comment by zhiweij1** [comment]
The failed cases are not related to this PR. They will be fixed in https://github.com/oneapi-src/SYCLomatic-test/pull/377

File: clang/test/dpct/helper_files_ref/include/lapack_utils.hpp [hunk 3]
```diff
@@ -1644,6 +1727,181 @@ inline int syhegvd(sycl::queue &q, int itype, oneapi::mkl::job jobz,
...
+/// Computes the size of workspace memory of syev/heev function.
+/// \return Returns 0 if no synchronous exception, otherwise returns 1.
+/// \param [in] q Device queue where computation will be performed. It must
+/// have the in_order property when using the USM mode (DPCT_USM_LEVEL_NONE is
+/// not defined).
+/// \param [in] jobz Must be job::novec or job::vec.
+/// \param [in] uplo Must be uplo::upper or uplo::lower.
+/// \param [in] n The order of the matrix A.
+/// \param [in] lda The leading dimension of the matrix A.
+/// \param [in] device_ws_size The device workspace size as a number of
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/lapack_utils.hpp]:
```suggestion
/// \param [out] device_ws_size The device workspace size as a number of
```

File: clang/test/dpct/helper_files_ref/include/lapack_utils.hpp [hunk 4]
```diff
@@ -1644,6 +1727,181 @@ inline int syhegvd(sycl::queue &q, int itype, oneapi::mkl::job jobz,
...
+/// \param [in] q Device queue where computation will be performed. It must
+/// have the in_order property when using the USM mode (DPCT_USM_LEVEL_NONE is
+/// not defined).
+/// \param [in] jobz Must be job::novec or job::vec.
+/// \param [in] uplo Must be uplo::upper or uplo::lower.
+/// \param [in] n The order of the matrix A.
+/// \param [in] a_type The data type of the matrix A.
+/// \param [in] lda The leading dimension of the matrix A.
+/// \param [in] w_type The data type of the eigenvalues.
+/// \param [in] device_ws_size The device workspace size as a number of
```
Comment (zhimingwang36) [inline clang/test/dpct/helper_files_ref/include/lapack_utils.hpp]:
```suggestion
/// \param [out] device_ws_size The device workspace size as a number of
```

File: clang/test/dpct/helper_files_ref/include/lapack_utils.hpp (hunk 3 above)
Comment (zhiweij1) [reply clang/test/dpct/helper_files_ref/include/lapack_utils.hpp]: fixed

File: clang/test/dpct/helper_files_ref/include/lapack_utils.hpp (hunk 4 above)
Comment (zhiweij1) [reply clang/test/dpct/helper_files_ref/include/lapack_utils.hpp]: fixed

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**